Enable PROXY protocol for specific CIDRs in HAProxy

Dariquest opened the pull request:
This PR introduces a new property **expect_proxy_cidrs**, which accepts a list of CIDR ranges for which to expect the PROXY protocol. This property allows selective enablement of PROXY protocol based on the source IP address.
**Expect_proxy_cidrs** is mutually exclusive with the **accept_proxy**, which enables PROXY protocol for all connections, and will lead to validation failure if both are set to true.

peanball commented on jobs/haproxy/spec:
```diff
@@ -553,6 +553,9 @@ properties:
     default: false
+  ha_proxy.expect_proxy:
+    description: "Turned off by default. Enables proxy protocol for the specified IP CIDR ranges. Only applies if `ha_proxy.accept_proxy` is disabled."
+    default: false
```
this should be a list of the CIDRs that should use the `expect-proxy` directive in HAProxy, not a boolean.

Dariquest replied:
thanks

peanball commented on jobs/haproxy/spec:
```diff
@@ -553,6 +554,9 @@ properties:
     default: false
+  ha_proxy.expect_proxy:
+    description: "Enables proxy protocol for the specified list of CIDR ranges. Only applies if `ha_proxy.accept_proxy` is disabled."
+    default: ""
```
if you put `default: ~`, it will not be set by default. The `spec` file is used to set default values in the manifest, if they're absent.

For the logic, I really you'd check for a list, and if `expect_proxy` is `nil` (because the value is not set), then you know that it's not active. This is particularly important for the mutual exclusivity with the `accept_proxy` property.

peanball commented on jobs/haproxy/spec:
```diff
@@ -580,6 +584,13 @@ properties:
       cidr_whitelist:
       - 172.168.4.1/32
       - 10.2.0.0/16
+    ha_proxy.cidr_proxies:
```
this list should just be the value of `expect_proxy`, not a separate property.

Dariquest replied:
done

peanball commented on jobs/haproxy/templates/haproxy.config.erb:
```diff
@@ -169,6 +169,9 @@ end
 # }}}
 
 # Error checking
+  if !p("ha_proxy.accept_proxy", true) && p("ha_proxy.expect_proxy", true)
```
`expect_proxy` should be the list, not a boolean.

Dariquest replied:
done

peanball commented on spec/haproxy/templates/proxies_cidrs.txt_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+          # BEGIN proxies cidrs
+          10.0.1.32/27
+          2600:1f18:4c0b:3207::/64
```
for IPv6 please use IP addresses from the documentation space:
* `2001:db8::/32`
* `3fff::/20`

See [RFC 9637](https://www.rfc-editor.org/rfc/rfc9637.html)

Dariquest replied:
thx

peanball reviewed (commented):
some WIP comments. I know this is not ready yet.

peanball commented on acceptance-tests/proxy_protocol_test.go:
```diff
@@ -61,9 +60,51 @@ var _ = Describe("Proxy Protocol", func() {
+		}, time.Minute, time.Second).Should(Equal("running"))
+
+		By("Sending a request with Proxy Protocol Header to HAProxy traffic port")
+		err := performProxyProtocolRequest(haproxyInfo.PublicIP, 80, "/")
```
in addition, we could have a test here that a regular (i.e. non-proxy-protocol) request fails, because HAProxy was in fact expecting proxy protocol.

Dariquest replied:
@peanball Thanks

Dariquest replied:
done

peanball commented on jobs/haproxy/templates/haproxy.config.erb:
```diff
@@ -169,6 +169,9 @@ end
 # }}}
 
 # Error checking
+  if !p("ha_proxy.accept_proxy", true) && !p("ha_proxy.expect_proxy", nil)
```
there should be a spect test for this in the haproxy.config frontend tests, to ensure that the flags are mutually exclusive.

Dariquest replied:
Thanks, will do.

Dariquest replied:
done

Dariquest commented on acceptance-tests/proxy_protocol_test.go:
```diff
@@ -61,9 +60,51 @@ var _ = Describe("Proxy Protocol", func() {
+  path: /instance_groups/name=haproxy/jobs/name=haproxy/properties/ha_proxy/expect_proxy?
+  value:
+    - 127.0.0.1/8
+    - ::1/128
```
@peanball Of course, we are not ready with the tests at all. We have only one without any meaning at first and still trying to execute it. The tests are not ready for review at all. We just pushed this stand because it was required by the tests.

peanball commented on jobs/haproxy/templates/haproxy.config.erb:
```diff
@@ -528,6 +531,9 @@ frontend https-in
       <%- end -%>
     <%- end -%>
   <%- end -%>
+  <% if p("ha_proxy.expect_proxy") -%>
```
add a check in the frontend spec test that this config line is output when you have the expected configuration.

peanball commented on jobs/haproxy/templates/haproxy.config.erb:
```diff
@@ -691,6 +697,9 @@ frontend wss-in
   <%- if properties.ha_proxy.frontend_config -%>
     <%= format_indented_multiline_config(p("ha_proxy.frontend_config")) %>
   <%- end -%>
+  <% if p("ha_proxy.expect_proxy") -%>
```
add a check in the frontend spec test that this config line is output when you have the expected configuration.

peanball commented on jobs/haproxy/templates/proxies_cidrs.txt.erb:
```diff
@@ -0,0 +1,21 @@
+# generated from proxies_cidrs.txt.erb
+<%
+require "base64"
```
these dependencies are not used. You based this file on logic that supports compressed files. Likely we can make this simpler?

A `join` would help, e..g `cidrs.join("\n")`.

maxmoehl reviewed (approved):
Approving for CI.

Dariquest commented:
> Approving for CI.

Thanks!

b1tamara reviewed (approved):
Approving for CI

b1tamara commented on acceptance-tests/proxy_protocol_test.go:
```diff
@@ -23,44 +25,88 @@ var _ = Describe("Proxy Protocol", func() {
+			err = performProxyProtocolRequest(haproxyInfo.PublicIP, 80, "/")
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Expecting Proxy Protocol for requests from IPs in the list")
```
I cannot find the request and the "Expect" check for this test case.

a18e replied:
see my comment below

b1tamara commented on acceptance-tests/proxy_protocol_test.go:
```diff
@@ -23,44 +25,88 @@ var _ = Describe("Proxy Protocol", func() {
+			expectConnectionResetErr(err)
+
+			err = performProxyProtocolRequest(haproxyInfo.PublicIP, 80, "/")
+			Expect(err).NotTo(HaveOccurred())
```
I am confused about this test case. If I understand correctly the test description, the test checks: no proxy protocol header -> no err, if proxy protocol header is set -> err. But the test is the other way around

a18e replied:
Good point, I cleaned up the comments for the test in https://github.com/sap-contributions/haproxy-boshrelease/commit/ffa1156f39e16e91a21366cc788f184abbe8ff76. Feel free to review again.

peanball replied:
it's good and consistent now.

b1tamara commented on jobs/haproxy/spec:
```diff
@@ -580,6 +581,13 @@ properties:
       cidr_whitelist:
       - 172.168.4.1/32
       - 10.2.0.0/16
+  ha_proxy.expect_proxy:
```
The property name is for me too general. I would expect boolean value for such property. Maybe use`expect_proxy_cidrs`. It is more clear.

Dariquest replied:
Good suggestion. It was discussed beforehand and the name was selected to ensure the similarity to accept_proxy. I would change it, but need one more opinion about it @peanball.

a18e replied:
I'm with @peanball on this one. I think we should keep it as is. `accept_proxy` is boolelan, `expect_proxy_cidrs` ist a list of CIDRS

Dariquest replied:
Rename done

maxmoehl reviewed (approved):
Approving for CI.

peanball reviewed (approved):
LGTM. The description, or commit message, could be improved.
@Dariquest received some feedback already on this. This PR os not about IPv6, but about conditional Proxy Protocol handling. It benefits dual-stack roll-out on AWS, but is not the only possible use.

Dariquest commented:
> LGTM. The description, or commit message, could be improved. @Dariquest received some feedback already on this. This PR os not about IPv6, but about conditional Proxy Protocol handling. It benefits dual-stack roll-out on AWS, but is not the only possible use.

Removed the IPv6 and adjusted the description. Thanks!

b1tamara commented on spec/haproxy/templates/haproxy_config/frontend_http_spec.rb:
```diff
@@ -42,6 +42,30 @@
+    context 'when ha_proxy.expect_proxy_cidrs is not empty/nil and ha_proxy.accept_proxy is false' do
+      let(:properties) do
+        { 'accept_proxy' => false,
+          'expect_proxy_cidrs' => ['127.0.0.1/8'] }
```
I would add an additional spec test to check that if the expect_proxy_cidrs is not explicitly set, then the line `tcp-request connection expect-proxy layer4 if { src -f /var/vcap/jobs/haproxy/config/expect_proxy_cidrs.txt` is not here. To prevent regressions in the future.

Dariquest replied:
@b1tamara Thanks Tamara, feel free to take a look again.

b1tamara commented on acceptance-tests/proxy_protocol_test.go:
```diff
@@ -23,44 +25,90 @@ var _ = Describe("Proxy Protocol", func() {
+		})
 	})
 })
 
```
It might make sense to have an edge test case with empty except_proxy_cidr and check that the call without proxy will be successful.

b1tamara commented:
Hi @Dariquest , everything fine now. The only minor thing, you might want to add this sentence to spec `expect_proxy_cidrs` description: `expect_proxy_cidrs is mutually exclusive with the accept_proxy` and not only fail if these two properties are set at the same time.

Dariquest commented:
> Hi @Dariquest , everything fine now. The only minor thing, you might want to add this sentence to spec `expect_proxy_cidrs` description: `expect_proxy_cidrs is mutually exclusive with the accept_proxy` and not only fail if these two properties are set at the same time.

Done, thanks!

b1tamara reviewed (approved):
LGTM
